nservicebus/cancellation-and-catching-exceptions.md

```diff
@@ -0,0 +1,138 @@
+
+    // re-throw the exception to propagate the cancellation to the caller of the current method
+    throw;
+}
+catch (Exception ex)
+{
+    // this catch (if it is required) will now catch only exceptions which do NOT represent cancellation,
+    // so it does not require a filter
+}
+```
+
+## Helper methods
+
+If exception handling is widespread, it may be helpful to introduce an `IsCausedBy` extension method:
+
+```csharp
+public static bool IsCausedBy(this Exception ex, CancellationToken cancellationToken) =>
+    ex is OperationCanceledException && cancellationToken.IsCancellationRequested;
+```
+
+Using this method, the `catch` filters are much simpler in both cases:
+
+```c#
+try
+{
+    await foo.Bar(cancellationToken).ConfigureAwait(false);
+}
+catch (Exception ex) when (!ex.IsCausedBy(cancellationToken))
+{
+    // foo.Bar failed — take appropriate action, including re-throwing the exception if appropriate
+}
+```
+
+```c#
+try
+{
+    await foo.Bar(cancellationToken).ConfigureAwait(false);
+}
+catch (Exception ex) when (ex.IsCausedBy(cancellationToken))
+{
+    // foo.Bar was cancelled — take appropriate action
+
+    // re-throw the exception to propagate the cancellation to the caller of the current method
+    throw;
+}
+```
+
+Note that the second example is catching `Exception` rather than `OperationCanceledException` and the `IsCausedBy` method is filtering on the exception type instead of the `catch` clause itself. This results in one extra type comparison (`isinst`) in the resulting IL code, but the performance cost is negligible and the code is simpler to read.
+
+## Inside the message processing pipeline
+
+For code inside the message processing pipeline, such as a message handler or saga, the above considerations are the same. The only difference is that the `CancellationToken` is provided by the `context.CancellationToken` property. For example:
+
+```c#
+public Task Handle(MyMessage, IMessageHandlerContext context)
+{
+    try
+    {
+        await foo.Bar(context.CancellationToken).ConfigureAwait(false);
+    }
+    catch (Exception ex) when (!ex.IsCausedBy(context.CancellationToken))
+    {
+        // foo.Bar failed — take appropriate action
+
+        // most of the time, the exception should be re-thrown,
+        // or a new exception should be thrown with the original exception as an inner exception:
+        //   throw new Exception("My message", ex);
+        // so that the message enters recoverability
+        throw;
+    }
+}
+```
+
+```c#
+public Task Handle(MyMessage, IMessageHandlerContext context)
+{
+    try
+    {
+        await foo.Bar(context.CancellationToken).ConfigureAwait(false);
+    }
+    catch (Exception ex) when (ex.IsCausedBy(context.CancellationToken))
+    {
+        // foo.Bar was cancelled — take appropriate action
+
+        // re-throw the exception to propagate the cancellation to the caller of the current method
+        throw;
+    }
+}
+```
```

Comment:
Given this is not in the menu and mostly for users of Particular.Analyzers, and we don't want users to catch exceptions in handlers anyway, these examples are just red herrings.
```suggestion
```